psycopg_c/psycopg_c/_psycopg/copy.pyx

```diff
@@ -23,13 +24,8 @@ cdef int32_t _binary_null = -1
 
 cdef object _format_row_binary(object row, Transformer tx, bytearray out):
     """Convert a row of adapted data to the data to send for binary copy"""
-    cdef Py_ssize_t rowlen
-    if type(row) is list:
-        rowlen = PyList_GET_SIZE(row)
-    elif type(row) is tuple:
-        rowlen = PyTuple_GET_SIZE(row)
-    else:
-        rowlen = len(row)
+    cdef row_fast = PySequence_Fast(row, "'row' is not a valid sequence")
```

Comment:
This might raise a TypeError if the row data is not a sequence or iterable. My question here would be, if this might raise on types, that worked before with the python level indexed access, e.g. do numpy arrays still work here?

The fact that it now also supports iterables out-of-the-box it a win API-wise.